### PR TITLE
[aoti] Prototype loading from bytes

### PR DESCRIPTION
**Pull request by angelayi**
Loader needs to have an official solution -- I'm pretty sure miniz can do this out of box, but haven't gotten the time to look at it yet. For now it just loads the buffer into a file.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/inductor/test_aot_inductor_package.py [hunk 1]
```diff
@@ -299,8 +300,12 @@ def forward(self, a, b):
             loaded1 = load_package(package_path, "model1")
             loaded2 = load_package(package_path, "model2")
 
-        assert torch.allclose(loaded1(*example_inputs1), ep1.module()(*example_inputs1))
-        assert torch.allclose(loaded2(*example_inputs2), ep2.module()(*example_inputs2))
+        self.assertTrue(
+            torch.allclose(loaded1(*example_inputs1), ep1.module()(*example_inputs1))
+        )
+        self.assertTrue(
+            torch.allclose(loaded2(*example_inputs2), ep2.module()(*example_inputs2))
```
Comment (henrylhtsang) [inline test/inductor/test_aot_inductor_package.py:307]: nit `torch.testing.assert_close`?

File: torch/_inductor/package/package.py [hunk 2]
```diff
@@ -241,9 +248,19 @@ def get_metadata(self) -> Dict[str, str]:
...
-def load_package(path: str, model_name: str = "model") -> AOTICompiledModel:  # type: ignore[type-arg]
-    if not path.endswith(".pt2"):
-        raise RuntimeError("Unable to load package. Path must be a .pt2 file.")
+def load_package(path: Union[str, io.BytesIO], model_name: str = "model") -> AOTICompiledModel:  # type: ignore[type-arg]
+    assert isinstance(path, io.BytesIO) or (
+        isinstance(path, str) and path.endswith(".pt2")
+    ), f"Unable to load package. Path must be a buffer or a file ending in .pt2. Instead got {path}"
+
+    if isinstance(path, io.BytesIO):
+        with tempfile.NamedTemporaryFile(suffix=".pt2") as f:
```
Comment (henrylhtsang) [inline torch/_inductor/package/package.py:257]: just to make sure, this will get deleted after we call the loader, and its okay right?

Comment (angelayi) [reply torch/_inductor/package/package.py:257]: yeah, because we should have an in-memory Runner

**Review by henrylhtsang — APPROVED** [review]

**Comment by henrylhtsang** [comment]
LOLLL so this actually works

**Pull request closed without merging**